samples/python_interop/cirq_samples/classical_control_and_barriers.ipynb

```diff
@@ -0,0 +1,213 @@
+    "    def _num_qubits_(self): return 1\n",
+    "    def _decompose_(self, qs):\n",
+    "        (bob,) = qs\n",
+    "        yield cirq.Z(bob).with_classical_controls(self.b1)\n",
+    "        yield cirq.X(bob).with_classical_controls(self.b2)\n",
+    "    def _circuit_diagram_info_(self, args): return f'Z^{self.b1}X^{self.b2}'\n",
+    "\n",
+    "# --- Optional full-qubit barrier (visual layer separator) ---\n",
+    "class BarrierGate(cirq.Gate):\n",
+    "    def __init__(self, qubits: Sequence[cirq.Qid]): self._n = len(qubits)\n",
+    "    def _num_qubits_(self): return self._n\n",
+    "    def _decompose_(self, qs): return []  # no-op\n",
+    "    def _circuit_diagram_info_(self, args):\n",
+    "        return cirq.CircuitDiagramInfo(wire_symbols=('│',) * self._n)\n",
+    "\n",
+    "# Teleportation ops as a flat list (auto packing may merge corrections upward if not separated)\n",
+    "def teleport_ops_with_optional_barrier(use_barrier: bool) -> Sequence[cirq.Operation]:\n",
+    "    alice = cirq.NamedQubit('alice')\n",
+    "    bob = cirq.NamedQubit('bob')\n",
+    "    msg = cirq.NamedQubit('msg')\n",
+    "    ops: Sequence[cirq.Operation] = []\n",
+    "    ops.append(BP.on(alice, bob))\n",
+    "    ops.append(cirq.rx(0.7)(msg))\n",
+    "    ops.append(MPrep.on(msg, alice))\n",
+    "    ops.append(cirq.measure(msg, key='b1'))\n",
+    "    ops.append(cirq.measure(alice, key='b2'))\n",
+    "    if use_barrier:\n",
+    "        ops.append(BarrierGate([msg, alice, bob]).on(msg, alice, bob))\n",
+    "    ops.append(MessageCorrection('b1','b2').on(bob))\n",
+    "    ops.append(cirq.measure(bob, key='result'))\n",
+    "    return ops\n",
+    "\n",
+    "# Explicit-moment construction for guaranteed boundaries\n",
+    "def teleport_explicit_moments() -> cirq.Circuit:\n",
+    "    alice = cirq.NamedQubit('alice')\n",
+    "    bob = cirq.NamedQubit('bob')\n",
+    "    msg = cirq.NamedQubit('msg')\n",
+    "    return cirq.Circuit([\n",
+    "        cirq.Moment([cirq.H(alice)]),\n",
+    "        cirq.Moment([cirq.CNOT(alice, bob), cirq.rx(0.7)(msg)]),\n",
+    "        cirq.Moment([cirq.CNOT(msg, alice)]),\n",
+    "        cirq.Moment([cirq.H(msg)]),\n",
+    "        cirq.Moment([cirq.measure(msg, key='b1'), cirq.measure(alice, key='b2')]),\n",
+    "        cirq.Moment([cirq.Z(bob).with_classical_controls('b1')]),\n",
+    "        cirq.Moment([cirq.X(bob).with_classical_controls('b2')]),\n",
+    "        cirq.Moment([cirq.measure(bob, key='result')]),\n",
+    "    ])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "699db40e",
+   "metadata": {},
+   "source": [
+    "## 3. Auto-Packed vs. Barrier vs. Explicit Moments\n",
+    "We construct three circuits:\n",
+    "1. Auto-packed (no barrier): may merge feed-forward corrections earlier.\n",
+    "2. Auto-packed with barrier: forces a visible layer boundary.\n",
+    "3. Explicit moments: manually defined phases, independent of greedy packing."
```

Comment:
I believe using explicit moments is the recommended approach. Barrier gate is a hack. Also the first circuit is simply incorrect in all senses.